### ISSUE TITLE
[REG]: colormap with symlog norm is broken by recent patch upstream (numpy)

### ISSUE THREAD

**Issue by neutrinoceros** [matplotlib/matplotlib#20843]
### Bug summary

Originally discovered while testing yt against the bleeding edge versions of matplotlib + numpy, and reported here https://github.com/yt-project/yt/issues/3472

As mentionned therein, the regression is caused by a recent patch merged into numpy's main branch https://github.com/numpy/numpy/pull/19244

After some digging, I was able to reproduce the error with a pure matplotlib test

### Code for reproduction

```python
from matplotlib import pyplot as plt
from matplotlib.colors import SymLogNorm
import numpy as np

def test_symlog_colorbar_simple():
    data = np.ones((5, 5))
    fig, ax = plt.subplots()
    cax = fig.add_axes((0.8, 0.1, 0.03, 0.9))
    norm = SymLogNorm(0.1, base=10, vmin=0, vmax=1)
    im = ax.imshow(data, norm=norm)
    fig.colorbar(im, cax)
```
running this with pytest

### Actual outcome

```
>       fig.colorbar(im, cax)

matplotlib/lib/matplotlib/figure.py:1156: in colorbar
    cb = cbar.Colorbar(cax, mappable, **cb_kw)
matplotlib/lib/matplotlib/colorbar.py:500: in __init__
    self.draw_all()
matplotlib/lib/matplotlib/colorbar.py:583: in draw_all
    self._add_solids(X, Y, self._values[ind, np.newaxis])
matplotlib/lib/matplotlib/colorbar.py:599: in _add_solids
    self.solids = self.ax.pcolormesh(
matplotlib/lib/matplotlib/__init__.py:1409: in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
matplotlib/lib/matplotlib/axes/_axes.py:6077: in pcolormesh
    X, Y, C, shading = self._pcolorargs('pcolormesh', *args,

    ...

E   ValueError: x and y arguments to pcolormesh cannot have non-finite values or be of type numpy.ma.core.MaskedArray with masked values
```

### Expected outcome

The same as against numpy 1.21.1 (current latest release): all green.
Further note that the same tes runs fine against numpy dev + matplotlib 3.4.3. It's really just a future collision between numpy 1.22.0 and matplotlib 3.5.0, so I expect it is still time to fix it, though I admit I am uncertain wether this should be considered a matplotlib bug or a numpy bug.

### Operating system

Running on OsX and ubuntu 18

### Matplotlib Version

'3.4.2.post1690+g3a4fdea8d2'

### Matplotlib Backend

'MacOSX'

### Python version

3.10.0b2

### Installation

source

**Comment by neutrinoceros** [comment]
@jklymak, unless I'm mistaken, I believe you have a lot of expertise with the colorbar setup code. There's no rush though, I just wanted to make sure this didn't fall off your radar :)

**Comment by jklymak** [comment]
Can you explain what a symlognorm with vmin of 0 and vmax of 1 is supposed to look like?  This seems like an ill formed norm and maybe the norm needs to do better error checking?

**Comment by neutrinoceros** [comment]
I'll admit that this specific parameter set is an oddball, though I would still expect it to be viable (though not really relevant in any situation I can think of).
Note that the bug can also be reproduced with
`SymLogNorm(0.1, base=10, vmin=-100, vmax=+100)`

**Comment by neutrinoceros** [comment]
I ran a few more examples around this, it seems like the test fails only when `vmin < data.min()` (`vmax > data.max()` isn't a problem though)

**Comment by jklymak** [comment]
I don't particularly have insight into why this would be failing.  I have quasi religious objections to symlognorm 😀 so someone who uses it should try and track down the problem.  I'm particularly confused by your symptoms, given that the data should not affect the colorbar.  Though maybe some of the norm changes @greglucas has been looking at are relevant.

**Comment by jklymak** [comment]
We do
```
def transform_non_affine(self, a):
        abs_a = np.abs(a)
        with np.errstate(divide="ignore", invalid="ignore"):
            out = np.sign(a) * self.linthresh * (
                self._linscale_adj +
                np.log(abs_a / self.linthresh) / self._log_base)
            inside = abs_a <= self.linthresh
        out[inside] = a[inside] * self._linscale_adj
```

However, at this point `out` has already been masked in the inner loop and the mask persists, even when `out[inside]` is assigned.   

This interacts with @anntzer `make_norm_from_scale` in some funny way.  He must be coercing to masked array?

**Comment by anntzer** [comment]
Even with a released numpy, the example gives a fairly weird colorbar, that seems to be a separate bug?
![test](https://user-images.githubusercontent.com/1322974/130278484-ddc4a92a-515f-40b8-b5cb-b045b69bf9be.png)

**Comment by neutrinoceros** [comment]
interesting, this could be the source a known bug in yt https://github.com/yt-project/yt/issues/3229, but yeah, that seems like an entirely separate bug to me

**Comment by anntzer** [comment]
I'll let you report it, then.

**Comment by neutrinoceros** [comment]
fair enough, will do :)

**Comment by anntzer** [comment]
I'm not convinced that numpy's new masking semantics make much sense (if you're assigning another *masked* array then I'd think you should copy the mask of the assigned array too?) but in case things stay as they are, the following patch seems to fix the problem:
```patch
diff --git i/lib/matplotlib/scale.py w/lib/matplotlib/scale.py
index 1809061067..04c1e985d8 100644
--- i/lib/matplotlib/scale.py
+++ w/lib/matplotlib/scale.py
@@ -369,6 +369,11 @@ class SymmetricalLogTransform(Transform):
                 np.log(abs_a / self.linthresh) / self._log_base)
             inside = abs_a <= self.linthresh
         out[inside] = a[inside] * self._linscale_adj
+        # If a is a maskedarray, out will have invalid values already masked,
+        # and the assignment to inside will not override the masking; i.e.
+        # these values must be manually unmasked.
+        if isinstance(a, np.ma.MaskedArray):
+            out.mask[inside] = np.ma.getmaskarray(a)[inside]
         return out
 
     def inverted(self):
```
Adding a test and turning this into a PR is left as an exercise to the reader.

**Comment by jklymak** [comment]
The weird colorbar is just because it is being placed manually.

**Comment by jklymak** [comment]
Btw I agree that numpy seems to have broken this by not copying the mask and assuming no mask if assigning a non_masked array.

**Comment by neutrinoceros** [comment]
> The weird colorbar is just because it is being placed manually.

ah ! my bad. It took me so long to boil this example down to pure mpl that I forgot about that. Indeed, not a bug.
Ok guys, thanks a lot for your feedback, I'll report to numpy directly.

**Comment by neutrinoceros** [comment]
There we go https://github.com/numpy/numpy/issues/19721

**Comment by jklymak** [comment]
I think you can show this to numpy more directly by creating a masked array, masking some elements, assigning them to a normal array and showing the mask persists.  (I'd provide an example but away from computer). A) I'm sure that behaviour changed and b) pretty sure it is undesirable.  I think you'll get more traction on the numpy issue if you leave matplotlib out of the report.

**Comment by neutrinoceros** [comment]
Good thinking, I'll double check that and update the report later (also AFK now).

**Comment by neutrinoceros** [comment]
fixed upstream https://github.com/numpy/numpy/pull/19736
I checked that the test in OP is indeed solved against matplotlib 3.4.3 and on your master branch
Thanks again @jklymak and @anntzer for you help with this, really appreciate it !
